ipcore/vhdl_node.py

```diff
@@ -0,0 +1,373 @@
+        """Generate VHDL assignment to this port, including necessary resizing.
+
+        Resizes using underlying data type if possible.
+        VHDL data types must match except for std_logic and std_logic_vector.
+        For assignmnents between std_logic and std_logic_vector,
+        the first bit is where data is assumed to be stored.
+
+        Parameters
+        ----------
+        new_signal_value : Signal
+            signal being assigned to this. VHDL data types must match except for std_logic and std_logic_vector
+
+        Returns
+        -------
+        str
+            VHDL code to assign new_signal_value to this port
+        """
+        return self.signal.generate_assignment(new_signal_value)
+
+class PortDir(Enum):
+    In = "in "
+    Out = "out"
+
+class ArraySignal(Signal):
+    def __init__(self, name, vhdl_type, length, element_type, element_length, element_data_type=None):
+        super().__init__(name, length, None, vhdl_type)
+        self.signals = [Signal(f"{self.name}({i})", element_length, None, element_type, element_data_type) for i in range(length)]
+    def __getitem__(self, key):
+        return self.signals[key]
+class Entity(BaseVHDLNode):
+    def __init__(self, name, ports):
+        super().__init__(name)
+        self.ports = ports
+    def generate(self):
+        """Generate a VHDL entity."""
+        vhdl_string = f"entity {self.name} is \n"
+        vhdl_string += self._generate_ports()
+        vhdl_string += f"end entity {self.name};\n\n"
+        return vhdl_string
+    def _generate_ports(self):
+        vhdl_string = ""
+        vhdl_string += tab() + "port (\n"
+        vhdl_string += tab(2)
+        ports_vhdl = [port.generate() for port in self.ports]
+        vhdl_string += (";\n" + tab(2)).join(ports_vhdl)
+        vhdl_string += "\n"
+        vhdl_string += tab() + ");\n"
+        return vhdl_string
+    def get_port(self, name):
+        """Get entity port by name.
+
+        Parameters
+        ----------
+        name : str
+            Name of port to retrieve
+
+        Returns
+        -------
+        Port, None
+            Returns entity port with the given name if found, otherwise None.
+        """
+        return next((port for port in self.ports if port.name == name), None)
+    def get_ports(self, *names):
+        """Get ports by name.
+
+        Returns
+        -------
+        list of Port
+            A list of Ports found matching the names given.
+        """
+        return filter(lambda port: port.name in names, self.ports)
+class Component(Entity):
+    def __init__(self, name, ports):
+        super().__init__(name, ports)
+        self.port_map = None
+    def generate(self):
+        """Generate a VHDL component."""
+        vhdl_string = f"component {self.name} is \n"
+        vhdl_string += self._generate_ports()
+        vhdl_string += f"end component {self.name};\n\n"
+        return vhdl_string
+class PortMap(BaseVHDLNode):
```

Comment:
Semantically, I feel like a port map is kind of a subclass of a component, as most of it is the same save for actually doing some assignments. 

If we all agree that a port map is semantically a subclass of a component, I'm not sure if making the class hierarchy reflect those semantics is possible or worthwhile. It's certainly not straight forward with the current class hierarchy, and I don't know if changing it would make the implementation any cleaner.

Reply:
I am not sure that really makes sense to do. If `PortMap` is just a subclass of `Component`, then you could pass a `PortMap` to an `Architecture` and it would generate a port map where a component should be.

Additionally, while not reflected properly in the code presently you only ever have one declaration of a particular component but you could have many port maps (as they are essentially component instances).

Reply:
Good point about it not being an actual subclass. We don't want that behavior. It makes more sense for a port map to contain a component.

My main thought was just looking for a way to avoid some repetition and more closely tie the port map to it's component. I don't have any better ideas than what is currently implemented at this point. The primary thing that caught my eye was making a dictionary for the port mapping and passing that into the `PortMap` constructor; it felt like you had basically already created the port map before actually creating a `PortMap` object. I don't know if there's really a cleaner way to do it than what you already have; at some level, you have to manually specify which signals get assigned to which ports in the port map.

Reply:
> The primary thing that caught my eye was making a dictionary for the port mapping and passing that into the PortMap constructor;
Fundamentally a port map without a mapping isn't a port map, hence `PortMap` requires the mapping in its constructor. The mapping in Python does happen consequently before I have made it, which is totally fine because the point of `PortMap` isn't to map ports to signals in Python, its to generate that code in VHDL. 

If it was a compiler and not a generator, inheriting from dictionary so you can natively do what I am doing with dictionaries with `PortMap`s instead would make sense.

Reply:
Thinking on it more, one potential option is `PortMap(component)` and `mapPort(portName, signal)`. The mapping would exist internally and default to None for everything (aka "open"). That way it guarantees all component ports have a mapping and that you can only map port's of the component rather than non-existent ones.

Reply:
While this isn't a massive change, given it is non-trivial and not strictly needed I am going to advocate for this just being made into an issue to be resolved soon so that we can get this merged in.